### PR TITLE
Type changes and fixing test cases.

### PR DESCRIPTION
**Pull request by rishi-aga**
Type changes and fixing test cases.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/metadata/MetaDataStoreTest.java [hunk 1]
```diff
@@ -20,22 +21,30 @@
...
-        dictionary.bindEntity(Player.class);
-        dictionary.bindEntity(CountryView.class);
-        dictionary.bindEntity(CountryViewNested.class);
+        dictionary.bindEntity(getType(PlayerStatsWithView.class));
+        dictionary.bindEntity(getType(PlayerStatsView.class));
+        dictionary.bindEntity(getType(PlayerStats.class));
+        dictionary.bindEntity(getType(Country.class));
+        dictionary.bindEntity(getType(SubCountry.class));
+        dictionary.bindEntity(getType(Player.class));
+        dictionary.bindEntity(getType(CountryView.class));
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/metadata/MetaDataStoreTest.java:46]: These getTypes are not necessary.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexTransaction.java [hunk 2]
```diff
@@ -200,43 +202,43 @@ public void updateToManyRelation(DataStoreTransaction relationTx,
...
     public Object getAttribute(Object entity, Attribute attribute, RequestScope scope) {
-        DataStoreTransaction transaction = getTransaction(entity.getClass());
+        DataStoreTransaction transaction = getTransaction(getType(entity.getClass()));
         return transaction.getAttribute(entity, attribute, scope);
     }
 
     @Override
     public void setAttribute(Object entity, Attribute attribute, RequestScope scope) {
-        DataStoreTransaction transaction = getTransaction(entity.getClass());
+        DataStoreTransaction transaction = getTransaction(getType(entity.getClass()));
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexTransaction.java:217]: Everywhere we call entity.getClass - we need to use the EntityDictionary.getTYpe instead.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java [hunk 3]
```diff
@@ -46,13 +48,13 @@ protected DataStoreTransaction beginTransaction(DataStore dataStore) {
     @Override
     public void save(Object entity, RequestScope requestScope) {
         getTransaction(entity).save(entity, requestScope);
-        dirtyObjects.add(this.multiplexManager.getSubManager(entity.getClass()), entity);
+        dirtyObjects.add(this.multiplexManager.getSubManager(getType(entity.getClass())), entity);
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java:51]: Same comment with entity.getClass.  There should be no calls to getClass here.  Instead call getType from the entity dictionary on the entity itself.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java [hunk 4]
```diff
@@ -137,7 +139,7 @@ private Object cloneObject(Object object) {
             return null;
         }
 
-        Class<?> cls = multiplexManager.getDictionary().lookupBoundClass(object.getClass());
+        Type<?> cls = multiplexManager.getDictionary().lookupBoundClass(getType(object.getClass()));
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java:142]: Same comment.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/MapEntryContainer.java [hunk 5]
```diff
@@ -49,7 +50,7 @@ public Object processFetch(Environment context, PersistentResourceFetcher fetche
                     + "'. Maps only contain fields 'key' and 'value'");
         }
 
-        if (nonEntityDictionary.hasBinding(returnObject.getClass())) {
+        if (nonEntityDictionary.hasBinding(new ClassType(returnObject.getClass()))) {
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/MapEntryContainer.java:53]: We should never call object.getClass directly anywhere in the code.  Always use the getType method in entity dictionary.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/NodeContainer.java [hunk 6]
```diff
@@ -43,7 +45,7 @@ public Object processFetch(Environment context, PersistentResourceFetcher fetche
                     .getAttributeMap().getOrDefault(context.field.getSourceLocation(), null);
             Object attribute = context.parentResource.getAttribute(requested);
 
-            if (attribute != null && nonEntityDictionary.hasBinding(attribute.getClass())) {
+            if (attribute != null && nonEntityDictionary.hasBinding(new ClassType(attribute.getClass()))) {
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/NodeContainer.java:48]: Never call getClass directly.  Use dictinoary.getType

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/NonEntityContainer.java [hunk 7]
```diff
@@ -35,7 +37,7 @@ public Object processFetch(Environment context, PersistentResourceFetcher fetche
             return null;
         }
 
-        if (nonEntityDictionary.hasBinding(object.getClass())) {
+        if (nonEntityDictionary.hasBinding(new ClassType(object.getClass()))) {
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/containers/NonEntityContainer.java:40]: Never call object.getClass directly.  Use dictionary.getType

File: elide-graphql/src/test/java/com/yahoo/elide/graphql/GraphQLTest.java [hunk 8]
```diff
@@ -29,10 +30,10 @@ public GraphQLTest() {
 
         dictionary = new EntityDictionary(checks);
 
-        dictionary.bindEntity(Book.class);
-        dictionary.bindEntity(Author.class);
-        dictionary.bindEntity(Publisher.class);
-        dictionary.bindEntity(Pseudonym.class);
-        dictionary.bindEntity(Address.class);
+        dictionary.bindEntity(getType(Book.class));
```
Comment (aklish) [inline elide-graphql/src/test/java/com/yahoo/elide/graphql/GraphQLTest.java:33]: Not necessary.

File: elide-graphql/src/test/java/com/yahoo/elide/graphql/ModelBuilderTest.java [hunk 9]
```diff
@@ -86,9 +89,9 @@
     public ModelBuilderTest() {
         dictionary = new EntityDictionary(Collections.EMPTY_MAP);
 
-        dictionary.bindEntity(Book.class);
-        dictionary.bindEntity(Author.class);
-        dictionary.bindEntity(Publisher.class);
+        dictionary.bindEntity(getType(Book.class));
```
Comment (aklish) [inline elide-graphql/src/test/java/com/yahoo/elide/graphql/ModelBuilderTest.java:92]: Not necessary.

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/compile/ElideDynamicEntityCompiler.java [hunk 10]
```diff
@@ -62,14 +64,14 @@ public ClassLoader getClassLoader() {
      * @throws ClassNotFoundException
      */
     @SuppressWarnings({ "rawtypes", "unchecked" })
-    public Set<Class<?>> findAnnotatedClasses(Class annotationClass)
+    public Set<Type<?>> findAnnotatedClasses(Class annotationClass)
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/compile/ElideDynamicEntityCompiler.java:67]: I'm not sure we want this to work with Type directly or instead use Class here.

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidator.java [hunk 11]
```diff
@@ -127,7 +129,7 @@ private void initialize() {
 
         annotatedClasses.forEach(cls -> {
             if (cls.getAnnotation(Include.class) != null) {
-                dictionary.bindEntity(cls);
+                dictionary.bindEntity(getType(cls));
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidator.java:132]: Not necessary to call getType here.

File: elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/ElideAutoConfiguration.java [hunk 12]
```diff
@@ -176,7 +180,10 @@ public void inject(Object entity) {
 
         if (isAggregationStoreEnabled(settings) && isDynamicConfigEnabled(settings)) {
             ElideDynamicEntityCompiler compiler = dynamicCompiler.getIfAvailable();
-            Set<Class<?>> annotatedClass = compiler.findAnnotatedClasses(SecurityCheck.class);
+            Set<Class<?>> annotatedClass = compiler.findAnnotatedClasses(SecurityCheck.class).stream()
+                            .map(ClassType.class::cast)
```
Comment (aklish) [inline elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/ElideAutoConfiguration.java:184]: See earlier comments.

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java [hunk 13]
```diff
@@ -142,8 +144,8 @@ private void logContext(RelationshipOp operation, Environment environment) {
 
         GraphQLType parent = environment.parentType;
         if (log.isDebugEnabled()) {
-            log.debug("{} {} fields with parent {}<{}>",
-                    operation, requestedFields, EntityDictionary.getSimpleName(parent.getClass()), parent.getName());
+            log.debug("{} {} fields with parent {}<{}>", operation, requestedFields,
+                            EntityDictionary.getSimpleName(new ClassType(parent.getClass())), parent.getName());
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java:148]: This should be EntityDictionary.getType(parent)

File: elide-integration-tests/src/test/java/com/yahoo/elide/initialization/IntegrationTest.java [hunk 14]
```diff
@@ -70,6 +70,7 @@ protected IntegrationTest(final Class<? extends ResourceConfig> resourceConfig,
         try {
             this.server = setUpServer();
         } catch (Exception e) {
+            e.printStackTrace();
```
Comment (aklish) [inline elide-integration-tests/src/test/java/com/yahoo/elide/initialization/IntegrationTest.java:73]: Let's remove this.

File: elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java [hunk 15]
```diff
@@ -724,8 +726,9 @@ public Swagger build() {
 
         /* Create a Model for each Elide entity */
         Map<String, Model> models = new HashMap<>();
-        for (Class<?> clazz : allClasses) {
-            models.putAll(converters.readAll(clazz));
+        for (Type<?> clazz : allClasses) {
+            Preconditions.checkState(clazz instanceof ClassType);
+            models.putAll(converters.readAll(((ClassType) clazz).getCls()));
```
Comment (aklish) [inline elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java:731]: This is wrong I think.  We need swagger to work for non Class-types as well.

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java [hunk 16]
```diff
@@ -727,8 +729,15 @@ public Swagger build() {
...
-            Preconditions.checkState(clazz instanceof ClassType);
-            models.putAll(converters.readAll(((ClassType) clazz).getCls()));
+            if (clazz instanceof ClassType) {
+                models.putAll(converters.readAll(((ClassType) clazz).getCls()));
+            } else if (clazz instanceof Type) {
+                ModelConverterContextImpl context = new ModelConverterContextImpl(Arrays.asList(converter));
+                context.resolve(clazz);
+                models.putAll(context.getDefinedModels());
+            } else {
+                models.putAll(converters.readAll(clazz, null));
```
Comment (aklish) [inline elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java:739]: What is the 2nd argument here?

File: elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java [hunk 17]
```diff
@@ -727,8 +729,15 @@ public Swagger build() {
         /* Create a Model for each Elide entity */
         Map<String, Model> models = new HashMap<>();
         for (Type<?> clazz : allClasses) {
-            Preconditions.checkState(clazz instanceof ClassType);
-            models.putAll(converters.readAll(((ClassType) clazz).getCls()));
+            if (clazz instanceof ClassType) {
+                models.putAll(converters.readAll(((ClassType) clazz).getCls()));
+            } else if (clazz instanceof Type) {
+                ModelConverterContextImpl context = new ModelConverterContextImpl(Arrays.asList(converter));
```
Comment (aklish) [inline elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java:735]: Let's add some comments here why this path is different.

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java (hunk 16 above)
Comment (rishi-aga) [reply elide-swagger/src/main/java/com/yahoo/elide/swagger/SwaggerBuilder.java]: removed it. It was `com.fasterxml.jackson.annotation.JsonView`. Internally it uses `null` value if not passed.

**Pull request merged**